File: opcua-webservice-spring/tutorials/lesson1/src/main/java/org/opcfoundation/uawebservicedemo/models/employee/EmployeeRepositoryType.java
package org.opcfoundation.uawebservicedemo.models.employee;

import org.eclipse.milo.opcua.stack.core.StatusCodes;
import org.eclipse.milo.opcua.stack.core.UaRuntimeException;
import org.eclipse.milo.opcua.stack.core.types.builtin.LocalizedText;
import org.opcfoundation.uawebservicedemo.database.EmployeeDatabase;
import org.opcfoundation.uawebservicedemo.database.entity.Employee;
import org.opcfoundation.uawebservicedemo.database.mapper.EmployeeMapper;
import org.opcfoundation.uawebservicedemo.models.EmployeeTwinSpace;
import org.opcfoundation.webserver.digitaltwin.digitaltwin.DigitalTwinRepositoryType;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListRequest;
import org.opcfoundation.webserver.service.message.digitaltwin.GetDigitalTwinListResponse;
import org.opcfoundation.webserver.types.digitaltwin.DigitalTwinDescriptor;

import java.util.List;
import java.util.concurrent.CompletableFuture;

// Defining entry of employee list
public class EmployeeRepositoryType extends DigitalTwinRepositoryType {
    public EmployeeRepositoryType(EmployeeTwinSpace twinSpace) {
        super("EmployeeRepositoryType", new LocalizedText("EmployeeRepository"), twinSpace);
    }

    @Override
    public CompletableFuture<GetDigitalTwinListResponse> onGetDigitalTwinList(GetDigitalTwinListRequest request)
    {
        try
        {
            // Database query
            EmployeeMapper employeeMapper = EmployeeDatabase.getEmployeeMapper();
            List<Employee> employees = employeeMapper.getAllEmployee();

            // Return general information of employee
            GetDigitalTwinListResponse response = new GetDigitalTwinListResponse();
            EmployeeDigitalTwinType employeeDigitalTwinType = ((EmployeeTwinSpace)nodeManager).employeeDigitalTwinType;

            for (Employee item : employees)
            {
                DigitalTwinDescriptor descriptor = new DigitalTwinDescriptor(
                        item.getId().toString(),
                        new LocalizedText(item.getName()),
                        employeeDigitalTwinType);

                response.add(descriptor);
            }

            return CompletableFuture.completedFuture(response);

        } catch (Exception e) {
            throw new UaRuntimeException(StatusCodes.Bad_ResourceUnavailable);
        }
    }
}

File: opcua-webservice-node/examples/demoserver/src/models/employee/EmployeeRepositoryType.ts
import { StatusCodes } from "opcua-webapi";
import { UaError, UaLocalizedText, makeUaStatusCode } from "opcua-webapi-ts";
import {
    DigitalTwinDescriptor,
    DigitalTwinRepositoryType,
    GetDigitalTwinListRequest,
    GetDigitalTwinListResponse,
} from "opcua-webservice-node";

import { prisma } from "../../connectors/prismaClient";
import { EmployeeTwinSpace } from "../EmployeeTwinSpace";

export class EmployeeRepositoryType extends DigitalTwinRepositoryType {
    constructor(twinSpace: EmployeeTwinSpace) {
        super("EmployeeRepositoryType", new UaLocalizedText("EmployeeRepository"), twinSpace);
    }

    override async onGetDigitalTwinList(
        _request: GetDigitalTwinListRequest,
    ): Promise<GetDigitalTwinListResponse> {
        const employees = await prisma.employee.findMany({
            select: {
                ID: true,
                Name: true,
            },
            orderBy: {
                ID: "asc",
            },
        });

        const response = new GetDigitalTwinListResponse();

        for (const item of employees) {
            response.add(
                new DigitalTwinDescriptor(
                    item.ID.toString(),
                    new UaLocalizedText(item.Name),
                    EmployeeTwinSpace.EmployeeDigitalTwinType,
                ),
            );
        }

        return response;
    }
}